### PR TITLE
add IntervalSequence

### PR DESCRIPTION
**Pull request by decentralion**
An IntervalSequence is an opaque read-only array of intervals, with
runtime validation that the intervals are non-overlapping and also have
no gaps between them, and that every interval has non-negative duration.
This way we can use intervals and not worry that the interval may be
out-of-order or have gaps, etc.

Test plan: Unit tests added to verify that the interval sequences have
the expected properties. Also, `git grep -F '$ReadOnlyArray<Interval>'`
only has hits within the interval module itself.

### CODE REVIEW

File: src/core/interval.js [hunk 1]
```diff
@@ -18,6 +18,34 @@ export type Interval = {|
...
+      );
+    }
+    if (endTimeMs < startTimeMs) {
+      throw new Error(
+        `negative length interval: ${endTimeMs} < ${startTimeMs}`
+      );
+    }
+    lastEndTime = endTimeMs;
+  }
+  return intervals;
```
Comment (wchargin) [inline src/core/interval.js:46]: Unsound; this type-checks but fails:

```javascript
    it("resists attack", () => {
      const i0 = {startTimeMs: 0, endTimeMs: 100};
      const i1 = {startTimeMs: 100, endTimeMs: 200};
      const is = intervalSequence([i0, i1]);
      i1.startTimeMs = 300;
      intervalSequence((is: IntervalSequence)); // fails
    });
```

Deep-copy?

File: src/core/interval.test.js [hunk 2]
```diff
@@ -3,9 +3,70 @@
...
+      const i1 = {startTimeMs: 100, endTimeMs: 99};
+      const thunk = () => intervalSequence([i1, i0]);
+      expect(thunk).toThrowError("negative length interval: 99 < 100");
+    });
+    it("accepts a sequence with -Infinity and +Infinity", () => {
+      const i0 = {startTimeMs: -Infinity, endTimeMs: 100};
+      const i1 = {startTimeMs: 100, endTimeMs: +Infinity};
+      expect(intervalSequence([i0, i1])).toEqual([i0, i1]);
+    });
+    it("rejects a sequence with NaNs", () => {
```
Comment (wchargin) [inline src/core/interval.test.js:62]: NaNs can still slip through. This fails:

```javascript
    it("rejects a sequence with just one NaN interval", () => {
      const i0 = {startTimeMs: NaN, endTimeMs: NaN};
      const thunk = () => intervalSequence([i0]);
      expect(thunk).toThrowError("NaN");
    });
```

File: src/core/interval.js [hunk 3]
```diff
@@ -18,6 +18,34 @@ export type Interval = {|
   +endTimeMs: TimestampMs,
 |};
 
+/**
+ * An interval sequence is an array of intervals with the following guarantees:
+ * - Every interval has non-negative time spam (i.e. the end time is greater
```
[the remark this answers is not in the document]
Comment (decentralion) [reply src/core/interval.js:23]: Done.

File: src/core/interval.js [hunk 4]
```diff
@@ -18,6 +18,34 @@ export type Interval = {|
   +endTimeMs: TimestampMs,
 |};
 
+/**
+ * An interval sequence is an array of intervals with the following guarantees:
```
[the remark this answers is not in the document]
Comment (decentralion) [reply src/core/interval.js:22]: Done.

File: src/core/interval.js (hunk 1 above)
Comment (decentralion) [reply src/core/interval.js:46]: Fixed.

File: src/core/interval.test.js (hunk 2 above)
Comment (decentralion) [reply src/core/interval.test.js:62]: Fixed.

File: src/core/interval.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (decentralion) [reply src/core/interval.js:23]: Yeah, I think being more restrictive here makes sense.

File: src/core/interval.test.js [hunk 5]
```diff
@@ -3,9 +3,87 @@
...
+      const i1 = {startTimeMs: 100, endTimeMs: 200};
+      const thunk = () => intervalSequence([i0, i1]);
+      expect(thunk).toThrowError("NaN");
+    });
+    it("resists mutation of the intervals", () => {
+      const i0 = {startTimeMs: 0, endTimeMs: 100};
+      const i1 = {startTimeMs: 100, endTimeMs: 200};
+      const is = intervalSequence([i0, i1]);
+      i1.startTimeMs = 300;
+      intervalSequence(is);
```
Comment (wchargin) [inline src/core/interval.test.js:83]: Mind explicitly checking that `is[0]` still has valid values, rather
than just checking that it still validates? This should be sufficient,
but it’s less obvious to me, and it’s nice for tests to be obvious.

**Review by wchargin — APPROVED** [review]
Thanks!

File: src/core/interval.test.js (hunk 5 above)
Comment (decentralion) [reply src/core/interval.test.js:83]: Done.

**Comment by wchargin** [comment]
> the expected properties. Also, `git grep -F "$ReadOnlyArray<Interval>"`

I’ve updated this to `git grep -F '$ReadOnlyArray<Interval>'`; the
double-quoted version doesn’t do what you want in POSIX shells.

Also, it’s pleasing to see all the existing uses gaining this
validation, with no errors!

**Pull request merged**